lib/deimos/active_record_consume/batch_consumption.rb

```diff
@@ -114,50 +117,64 @@ def compact_messages(batch)
       # All messages are split into slices containing only unique keys, and
       # each slice is handles as its own batch.
       # @param messages [Array<Message>] List of messages.
-      # @return [void]
+      # @return [Array<Array<ActiveRecord::Base>, Array<BatchRecord>>]
       def uncompacted_update(messages)
-        BatchSlicer.
-          slice(messages).
-          each(&method(:update_database))
+        valid_records = []
+        invalid_records = []
+        BatchSlicer.slice(messages).each do |slice|
+          valid, invalid = update_database(slice)
+          valid_records.push(*valid) if valid.any?
+          invalid_records.push(*invalid) if invalid.any?
+        end
+        [valid_records, invalid_records]
       end
 
       # Perform database operations for a group of messages.
       # All messages with payloads are passed to upsert_records.
       # All tombstones messages are passed to remove_records.
       # @param messages [Array<Message>] List of messages.
-      # @return [void]
+      # @return [Array<Array<ActiveRecord::Base>, Array<BatchRecord>>]
       def update_database(messages)
         # Find all upserted records (i.e. that have a payload) and all
         # deleted record (no payload)
         removed, upserted = messages.partition(&:tombstone?)
 
         max_db_batch_size = self.class.config[:max_db_batch_size]
+        valid_upserts = []
+        invalid_upserts = []
         if upserted.any?
+          valid_upserts, invalid_upserts = if max_db_batch_size
+                                             upserted.each_slice(max_db_batch_size) do |group|
+                                               valid, invalid = upsert_records(group)
+                                               valid_upserts.push(*valid)
+                                               invalid_upserts.push(*invalid)
+                                             end
+                                             valid_upserts.compact!
+                                             invalid_upserts.compact!
```

Comment:
Is there a reason you're compacting inside each iteration instead of after all iterations are done?